### PR TITLE
[CmpInstAnalysis] Return decomposed bit test as struct (NFC)

### PR DESCRIPTION
**Pull request by nikic**
decomposeBitTestICmp() currently returns the result via two out parameters plus an in-place modification of Pred. This changes it to return an optional struct instead.

The motivation here is twofold. First, I'd like to extend this code to handle cases where the comparison is against a value other than zero, which would mean yet another out parameter. Second, while doing that I was badly bitten by the in-place modification, so I'd like to get rid of it.

### CODE REVIEW

File: llvm/lib/Transforms/Scalar/LoopIdiomRecognize.cpp [hunk 1]
```diff
@@ -2465,10 +2465,16 @@ static bool detectShiftUntilBitTestIdiom(Loop *CurLoop, Value *&BaseX,
   };
   auto MatchDecomposableConstantBitMask = [&]() {
     APInt Mask;
-    return llvm::decomposeBitTestICmp(CmpLHS, CmpRHS, Pred, CurrX, Mask) &&
-           ICmpInst::isEquality(Pred) && Mask.isPowerOf2() &&
-           (BitMask = ConstantInt::get(CurrX->getType(), Mask)) &&
-           (BitPos = ConstantInt::get(CurrX->getType(), Mask.logBase2()));
+    auto Res = llvm::decomposeBitTestICmp(CmpLHS, CmpRHS, Pred);
+    if (Res && Res->Mask.isPowerOf2()) {
+      assert(ICmpInst::isEquality(Res->Pred));
```
Comment (goldsteinn) [inline llvm/lib/Transforms/Scalar/LoopIdiomRecognize.cpp:2470]: Why did this become an assert?

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp [hunk 2]
```diff
@@ -870,11 +872,15 @@ static Value *foldSignedTruncationCheck(ICmpInst *ICmp0, ICmpInst *ICmp1,
                            APInt &UnsetBitsMask) -> bool {
     CmpInst::Predicate Pred = ICmp->getPredicate();
     // Can it be decomposed into  icmp eq (X & Mask), 0  ?
-    if (llvm::decomposeBitTestICmp(ICmp->getOperand(0), ICmp->getOperand(1),
-                                   Pred, X, UnsetBitsMask,
-                                   /*LookThroughTrunc=*/false) &&
-        Pred == ICmpInst::ICMP_EQ)
+    if (auto Res =
+            llvm::decomposeBitTestICmp(ICmp->getOperand(0), ICmp->getOperand(1),
+                                       Pred, /*LookThroughTrunc=*/false);
```
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:877]: Imo just yank the auto Res = llvm:::dec.... out of the if.

File: llvm/lib/Transforms/Scalar/LoopIdiomRecognize.cpp (hunk 1 above)
Comment (nikic) [reply llvm/lib/Transforms/Scalar/LoopIdiomRecognize.cpp:2470]: decomposeBitTestICmp() is required to always return an equality predicate.

**Review by goldsteinn — APPROVED** [review]
LGTM

File: llvm/lib/Transforms/Scalar/LoopIdiomRecognize.cpp [hunk 3]
```diff
@@ -2465,10 +2465,16 @@ static bool detectShiftUntilBitTestIdiom(Loop *CurLoop, Value *&BaseX,
   };
   auto MatchDecomposableConstantBitMask = [&]() {
     APInt Mask;
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/Scalar/LoopIdiomRecognize.cpp:2467]: It is unused now.

**Review by dtcxzyw — APPROVED** [review]
LG

**Pull request merged**